core/src/test/java/org/apache/calcite/test/ObjectArrayTableTest.java

```diff
@@ -0,0 +1,239 @@
+    void test(Connection connection) throws SQLException;
+  }
+
+  /**
+   * Execute a test with an in-memory table having java.sql.Date, Time, Timestamp columns.
+   * @param test the actual test code
+   */
+  private void withJavaSqlDateTypes(TestWithConnection test) throws SQLException {
+    try (
+        Connection connection = DriverManager.getConnection("jdbc:calcite:");
+        CalciteConnection calciteConnection = connection.unwrap(CalciteConnection.class)
+    ) {
+      final SchemaPlus rootSchema = calciteConnection.getRootSchema();
+
+      final JavaTypeFactoryImpl typeFactory =
+          new JavaTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+
+      final RelDataType rowType = typeFactory.createStructType(
+          Arrays.asList(
+              typeFactory.createJavaType(java.sql.Date.class),
+              typeFactory.createJavaType(java.sql.Time.class),
+              typeFactory.createJavaType(java.sql.Timestamp.class)
+          ),
+          Arrays.asList("dt", "tm", "ts")
+      );
+
+      final Enumerable<Object[]> enumerable =
+          Linq4j.asEnumerable(
+              Collections.singletonList(new Object[]{
+                  java.sql.Date.valueOf("2018-12-14"),
+                  java.sql.Time.valueOf("18:29:34"),
+                  java.sql.Timestamp.valueOf("2018-12-14 18:29:34.123")}));
+
+      final Table table = new AbstractQueryableTable(Object[].class) {
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        @Override public Queryable<Object[]> asQueryable(QueryProvider queryProvider,
+                                                         SchemaPlus schema, String tableName) {
+          return EnumerableDefaults.asOrderedQueryable(enumerable);
+        }
+
+        @Override public RelDataType getRowType(RelDataTypeFactory typeFactory) {
+          return rowType;
+        }
+      };
+
+      rootSchema.add("java_sql_date_types", table);
+
+      test.test(connection);
+    }
+  }
+
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-1703">[CALCITE-1703]
+   * Before the fix, functions on TIME and TIMESTAMP column can throw ClassCastException</a>. */
+  @Test public void testJavaSqlDateColumnReference() throws SQLException {
+
+    withJavaSqlDateTypes(connection -> {
+      // 'extract' function expects numeric date representation.
+      // RexToLixTranslator.convert() handles Java type to numeric storage type conversion.
+      // Storage types are defined at EnumUtils.toInternal.
+      final Statement statement = connection.createStatement();
+      String sql = "select extract(year from \"dt\"),"
+          + " extract(hour from \"tm\"),"
+          + " extract(day from \"ts\")"
+          + " from \"java_sql_date_types\"";
+
+      ResultSet resultSet = statement.executeQuery(sql);
+      assertTrue(resultSet.next());
+      assertEquals("java.sql.Date is converted to int when it's passed to a function",
```

Comment:
Thank you for adding the message, however:
1) "conversion to int" is a current implementation detail, so it might easily change
2) The test does NOT verify if `Date` is converted to int or not. The test verifies that end-to-end value is valid (it is good by the way!)

Current test would fail like

```
AssertionError: java.sql.Date is converted to int when it's passed to a function, got: 2017, expected: 2018
```

I don't think the message clarifies why 2018 was expected.

I suggest message should be like

`assertEquals("extract(year from Date.valueOf(\"2018-12-14\")", 2018, resultSet.getInt(1));`

In case of failure, it will fail like the following:

```
AssertionError: extract(year from Date.valueOf("2018-12-14"), got: 2017, expected: 2018
```

That enables to manually check if the expected value is indeed right.

Reply:
Thanks for clarifying the idea. That makes sense. I've updated the assertion messages.